packages/generator/src/name-formatting-util.ts

```diff
@@ -0,0 +1,138 @@
+/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
+import {
+  reservedObjectPrototypeKeywords,
+  reservedVdmKeywords
+} from './name-formatting-reserved-key-words';
+
+type Separator = '-' | '_';
+const MAXIMUM_NUMBER_OF_SUFFIX = 1000;
+
+export interface FindUniqueName {
+  nameToCheckForUniqueness: string;
+  relatedNamesBuilder?: (name: string) => string[];
+  alreadyUsedNames: string[];
+}
+
+export interface UniqueName {
+  uniqueName: string;
+  relatedUniqueNames: string[];
+}
+
+function getAllNames(name: FindUniqueName): string[] {
+  const allNames = [name.nameToCheckForUniqueness];
+  if (typeof name.relatedNamesBuilder !== 'undefined') {
+    allNames.push(...name.relatedNamesBuilder(name.nameToCheckForUniqueness));
+  }
+  return allNames;
+}
+
+function castToUniqueName(findUniqueName: FindUniqueName): UniqueName {
+  const name = findUniqueName.nameToCheckForUniqueness;
+  return {
+    uniqueName: name,
+    relatedUniqueNames: findUniqueName.relatedNamesBuilder
+      ? findUniqueName.relatedNamesBuilder(name)
+      : []
+  };
+}
+
+function isNameAndRelatedUnique(findUniqueName: FindUniqueName): boolean {
+  const allNames = getAllNames(findUniqueName);
+
+  return !allNames.some(name =>
+    isNameAlreadyUsed(name, findUniqueName.alreadyUsedNames)
+  );
+}
+
+function isNameAlreadyUsed(nameToTest: string, alreadyUsedNames: string[]) {
```

Comment:
I would generally try to refrain from giving names more context that they need. I think `name` is perfectly fine here.

Reply:
I think we can also improve the "already used" wording. If the second parameter is called `alreadyUsedNames` and the function is called `isNameAlreadyUsed` it implies that this function only checks `alreadyUsedNames.includes(nameToTest)`.

If we align to only use one word for something that cannot be taken anymore we could also call that `reservedNames` and `isReserved`.

Reply:
I included a class holding more state which makes the signatures simpler.